csrc/scheduler/cache_policy_refiner.cpp

```diff
@@ -0,0 +1,138 @@
+// clang-format off
+/*
+ * SPDX-FileCopyrightText: Copyright (c) 2023-present NVIDIA CORPORATION & AFFILIATES.
+ * All rights reserved.
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+// clang-format on
+
+#include <fusion.h>
+#include <ir/base_nodes.h>
+#include <ir/internal_nodes.h>
+#include <ir/utils.h>
+#include <root_domain_map.h>
+#include <scheduler/cache_policy_refiner.h>
+#include <scheduler/debug_utils.h>
+
+namespace nvfuser {
+
+namespace {
+
+// Returns whether a pointwise expression `expr` expands its input operand
+// `in_tv`.
+bool pointwiseExpands(const Expr* expr, const TensorView* in_tv) {
+  NVF_CHECK(
+      expr->outputs().size() == 1,
+      "A pointwise expression is expected to have one output: ",
+      expr->toString());
+  const Val* out = expr->output(0);
+
+  if (!out->isA<TensorView>()) {
+    return false;
+  }
+  const auto* out_tv = out->as<TensorView>();
+
+  auto root_domain_map =
+      PairwiseRootDomainMap(in_tv, out_tv)
+          .mapBroadcast(true)
+          .mapProducerToConsumer(in_tv->domain(), out_tv->domain());
+  return std::find_if(
+             root_domain_map.begin(),
+             root_domain_map.end(),
+             [](const auto& mapping) {
+               return mapping.first->isBroadcast() &&
+                   !mapping.second->isBroadcast();
+             }) != root_domain_map.end();
+}
+
+// Finds the first expanding use of `ldst`'s output, bypassing all pointwise
+// operations.
+const Expr* findExpand(const LoadStoreOp* ldst) {
+  std::queue<const Expr*> q;
+  std::unordered_set<const Expr*> visited;
+
+  auto enqueueIfNotVisited = [&q, &visited](const Expr* expr) {
+    if (visited.insert(expr).second) {
+      q.push(expr);
+    }
+  };
+
+  enqueueIfNotVisited(ldst);
+  while (!q.empty()) {
+    const Expr* def = q.front();
+    q.pop();
+    for (const Val* def_out : def->outputs()) {
+      if (!def_out->isA<TensorView>()) {
+        continue;
+      }
+
+      for (const Expr* use : def_out->uses()) {
+        if (use->isA<ExpandOp>()) {
+          return use;
+        }
+
+        if (use->isOneOf<UnaryOp, BinaryOp, TernaryOp, BroadcastOp>() ||
+            (use->isA<LoadStoreOp>() &&
+             use->as<LoadStoreOp>()->out()->as<TensorView>()->getMemoryType() !=
+                 MemoryType::Global)) {
```

Comment:
Is this to avoid applying the new policy when a tensor is just loaded and then immediately stored?

Even in that case, as long as there's redundant reads, i.e., `pointwiseExpands` returning `true`, wouldn't it make sense to use caching?

Reply:
My initial reason is to ensure the algorithm is linear. In a pathological case, we can have a chain of `ld.global -> st.global -> ld.global -> st.global -> ...`. If the algorithm doesn't stop at a `st.global`, it may traverse downstream exprs many times. 

> Even in that case, as long as there's redundant reads, i.e., pointwiseExpands returning true, wouldn't it make sense to use caching?

I think you are right. I need to figure out an easier way to cover such cases. In general, I'm inclined to keep optimizations simple unless there's a good (performance) reason to make them complicated.

Reply:
Makes sense. No need to make it so complicated at this moment. For simplicity, I originally thought it would be sufficient to look at reads of fusion inputs, but your approach also seems fine as well.